text/0000-extern-types.md

```diff
@@ -0,0 +1,128 @@
+- Feature Name: extern_types
+- Start Date: 2017-01-18
+- RFC PR: 
+- Rust Issue: 
+
+# Summary
+[summary]: #summary
+
+Add an `extern type` syntax for declaring types which are opaque to Rust's type
+system.
+
+# Motivation
+[motivation]: #motivation
+
+When interacting with external libraries we often need to be able to handle pointers to data that we don't know the size or layout of.
+
+In C it's possible to declare a type but not define it.
+These incomplete types can only be used behind pointers, a compilation error will result if the user tries to use them in such a way that the compiler would need to know their layout.
+
+In Rust, we don't have this feature. Instead, a couple of problematic hacks are used in its place.
+
+One is, we define the type as an uninhabited type. eg.
+
+```rust
+enum MyFfiType {}
+```
+
+Another is, we define the type with a private field and no methods to construct it.
+
+```rust
+struct MyFfiType {
+    _priv: (),
+}
+```
+
+The point of both these constructions is to prevent the user from being able to create or deal directly with instances of the type.
+Neither of these types accurately reflect the reality of the situation.
+The first definition is logically problematic as it defines a type which can never exist.
+This means that references to the type can also—logically—never exist and raw pointers to the type are guaranteed to be
```

Comment:
It won't break anything so long as you keep your uninhabited type behind a raw pointer, but as soon as you put it behind a reference you've got a situation which is statically impossible unless you're lying to the type system with `unsafe`.

This caused a breakage in the standard library implementation when I was implementing some of the uninhabitedness stuff. `&Void` was being used somewhere as a `void *` with a lifetime and (I think - my memory's a little vague) the compiler started assuming that a function returning `&Void` could never return, and so segfaults resulted.